### PR TITLE
DOC: add a brief note on "Protocols for methods" to NEP 18

### PR DESCRIPTION
**Pull request by shoyer**
Please take a look @hameerabbasi @mhvk @mrocklin @eric-wieser 

Once we merge this, I think we're ready to send it out to the list for final approval?

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Looks good to me.

**Comment by mhvk** [comment]
Yes, would be good to have that as another aspect that needs more thought.

File: doc/neps/nep-0018-array-function-protocol.rst [hunk 1]
```diff
@@ -799,6 +795,28 @@ If we want to include them in the future, the easiest way to do so would be to
 update the ``array_function_dispatch`` decorator to add them as function
 attributes.
 
+Protocols for methods
+~~~~~~~~~~~~~~~~~~~~~
+
+NumPy defines a number of classes with methods that act on NumPy arrays, e.g.,
+``vectorize.__call__`` and various ``random.RandomState`` methods. Such classes
```
Comment (eric-wieser) [inline doc/neps/nep-0018-array-function-protocol.rst]: I don't think `__call__` is a good example - I only think the first of these is reasonable:

```
__array_function__(some_vectorize_obj, ...)
__array_function__(some_vectorize_obj.__call__, ...)
__array_function__(some_vectorize_obj.__call__.__call__, ...)
```

Comment (mhvk) [reply doc/neps/nep-0018-array-function-protocol.rst]: Hmm, agree that the description can be seen as misleading. Certainly whatever is passed to `__array_function__` should be the "callable", i.e., the instance with a `__call__` method. In the context of the NEP, I guess the main question -- for which the answer is deferred -- is how to easily recognize those, as one cannot just, say, index a dict keyed with numpy functions with it, but will have to check the type or so.

Comment (shoyer) [reply doc/neps/nep-0018-array-function-protocol.rst]: What if I replace `vectorize.__call__` with `vectorize` here?

**Comment by shoyer** [comment]
> If you don't mind, we should include a similar section on array generation functions without explicitly proposing a solution (maybe mentioning the array_reference argument that I proposed earlier, along with maybe something like context managers, etc.)

My inclination would be to add something about this to NEP 22 instead. NEP 18 is focused on this specific protocol, but array generation functions (beyond those that already exist) falls under new APIs needed for high level array APIs.

File: doc/neps/nep-0018-array-function-protocol.rst (hunk 1 above)
Comment (eric-wieser) [reply doc/neps/nep-0018-array-function-protocol.rst]: Well then it won't really be a method, so I don't think this heading will make sense.

Is the case you're actually trying to cover here "callable objects generated at runtime" (methods close over self, vectorize closes over a lambda, etc), for which there is no global name to check against?

Comment (shoyer) [reply doc/neps/nep-0018-array-function-protocol.rst]: > In the context of the NEP, I guess the main question -- for which the answer is deferred -- is how to easily recognize those, as one cannot just, say, index a dict keyed with numpy functions with it, but will have to check the type or so.

This is a good way of putting it -- I'll add something about this

**Comment by hameerabbasi** [comment]
> My inclination would be to add something about this to NEP 22 instead. NEP 18 is focused on this specific protocol, but array generation functions (beyond those that already exist) falls under new APIs needed for high level array APIs.

I'd like to disagree here. A lot of code has been written using e.g. `np.ones(...)` or similar array generation functions, and mixing this code with other code from e.g. Dask may lead to out of memory errors, or producing arrays without units, and so on.

It would be nice to be able to say: "for this function/block of code, I'd like to follow a given implementation entirely or only for array generation functions, including calls to any upstream libraries". This is, in essence, an override.

However, I do not have time in the near term, but long term I'd love to work on a NEP especially for this sort of thing. Perhaps you're right that NEP-22 is the right place to put this, rather than this NEP.

**Comment by shoyer** [comment]
> It would be nice to be able to say: "for this function/block of code, I'd like to follow a given implementation entirely or only for array generation functions, including calls to any upstream libraries". This is, in essence, an override.

Do you think this would conflict with the `__array_function__` overrides envisioned here? If so, it may need to be addressed here...

**Comment by shoyer** [comment]
@hameerabbasi OK, but `np.arange()` would not use `__array_function__` at all in the current proposal, because it does not take any array-like arguments. Can you think of any functions that could use both?

**Comment by hameerabbasi** [comment]
Of course. I'm proposing that all current functions under `__array_function__` would be superseded by the backend implementation and fall back if (for example) `__array_backend__` returned `NotImplemented`.

**Comment by shoyer** [comment]
I've adjusted the text to refer more generally to "callable objects generated at runtime" as suggested by @eric-wieser.

I haven't mentioned anything about setting a "backend" per @hameerabbasi's suggestion, because I really don't see any interaction with the `__array_function__` protocol. Even if these take precedence over `__array_function__` overloads, they would not influence the way in which this protocol itself is designed.

**Comment by shoyer** [comment]
Any objections to merging this in its current state? I'd like to send out the official "this NEP will be marked as accepted in 7 days" notice soon...

**Comment by hameerabbasi** [comment]
None from me.

**Comment by mrocklin** [comment]
None from me

**Pull request merged**